### PR TITLE
Created the unit test for removing the warnings from the nurbs utilities [core]

### PR DESCRIPTION
**Pull request by andreas-apostolatos**
Dear all,

I am adding here a unit test to call functions from the nurbs utilities which are currently not used and cause the generation of warnings. I took this chance to also fix a buggy version of the projection function for the nurbs surface geometry, so please take time to also review that, it is not long at all.

Thanks!

### CODE REVIEW

**Comment by philbucher** [comment]
I think you have a memory error when running the tests according to Travis

File: kratos/tests/cpp_tests/geometries/test_projection_nurbs_geometry.cpp [hunk 1]
```diff
@@ -188,20 +188,21 @@ typedef Node<3> NodeType;
...
 
-         // Try projection with initial guess at u = 0.5, v = 0.5
+        // Try projection with initial guess at u = 0.5, v = 0.5
         parameter[0] = 0.5;
         parameter[1] = 0.5;
 
         // Try projection
-        bool isConverged2 = ProjectionNurbsGeometryUtilities::NewtonRaphsonSurface(parameter, point, projected_point, surface);
+        bool isConverged2 = ProjectionNurbsGeometryUtilities::NewtonRaphsonSurface(parameter, point, projected_point, surface, 100, 1e-12);
+        exit(-1);
```
Comment (jcotela) [inline kratos/tests/cpp_tests/geometries/test_projection_nurbs_geometry.cpp]: I do not think this should be here :confused:

Comment (andreas-apostolatos) [reply kratos/tests/cpp_tests/geometries/test_projection_nurbs_geometry.cpp]: Yes indeed! Thanks

**Comment by andreas-apostolatos** [comment]
> I think you have a memory error when running the tests according to Travis

I think it is due to the exit(-1) statement that @jcotela just found, thanks guys!

File: kratos/tests/cpp_tests/geometries/test_projection_nurbs_geometry.cpp [hunk 2]
```diff
@@ -188,20 +188,20 @@ typedef Node<3> NodeType;
         KRATOS_CHECK_NEAR(parameter[1], 0.5, TOLERANCE);
         KRATOS_CHECK_VECTOR_NEAR(projected_point, projected_point_expected1, TOLERANCE);
 
-        // Try projection of point (0.06, -0.06, 0.06) close to the singularity
+        // Try projection of point (0.1, -0.005, 0.005) close to the singularity
```
Comment (philbucher) [inline kratos/tests/cpp_tests/geometries/test_projection_nurbs_geometry.cpp:191]: you sure you didn't but a `0` too much?

File: kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+#include "containers/pointer_vector.h"
+#include "geometries/nurbs_shape_function_utilities/nurbs_utilities.h"
+
+#include "tests/cpp_tests/geometries/test_geometry.h"
+
+namespace Kratos {
+namespace Testing {
+
+        ///// Tests
+        KRATOS_TEST_CASE_IN_SUITE(removingWarningsNurbsUtilities, KratosCoreNurbsGeometriesFastSuite) {
```
Comment (philbucher) [inline kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp]:
```suggestion
        KRATOS_TEST_CASE_IN_SUITE(NurbsUtilities, KratosCoreNurbsGeometriesFastSuite) {
```

File: kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
+//    |  /           |
```
Comment (philbucher) [inline kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp]: please remove `removing_warnings` from the file name

File: kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            // Dummy number of control points
+            const int number_of_control_points = 2;
+            
+            // Dummy knot vector
+            Vector knot_vector = ZeroVector(number_of_knots);
+            knot_vector[0] = 0.0;
+            knot_vector[1] = 1.0;
+
+            // Dummy curve parameter
+            double parameter_t = .5;
```
Comment (philbucher) [inline kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp]:
```suggestion
            double parameter_t = 0.5;
```

File: kratos/utilities/nurbs_utilities/projection_nurbs_geometry_utilities.h [hunk 6]
```diff
@@ -124,15 +124,15 @@ namespace Kratos
         const double Accuracy = 1e-6)
     {
         // Initialize variables
-        bool clampU = false;
-        bool clampV = false;
+        bool conditionFirstRowZero, conditionSecondRowZero, conditionFirstColumnZero, conditionSecondColumnZero, isSystemInvertible;
```
Comment (philbucher) [inline kratos/utilities/nurbs_utilities/projection_nurbs_geometry_utilities.h]: I see that some of them are already there but the names should be lower case with underscores
Can you please update?

**Review by philbucher — COMMENTED** [review]
thx for fixing
I have some minor requests related to naming

also check Travis, now your test fails

File: kratos/utilities/nurbs_utilities/projection_nurbs_geometry_utilities.h (hunk 6 above)
Comment (andreas-apostolatos) [reply kratos/utilities/nurbs_utilities/projection_nurbs_geometry_utilities.h]: Any ideas why travis is still failing? It seems that it fails in some other tests .. Thanks!

Comment (philbucher) [reply kratos/utilities/nurbs_utilities/projection_nurbs_geometry_utilities.h]: `TestNurbsSurfaceGeometryProjection3d` fails

Comment (andreas-apostolatos) [reply kratos/utilities/nurbs_utilities/projection_nurbs_geometry_utilities.h]: Right, got to fix that, danke!

File: kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp (hunk 4 above)
Comment (philbucher) [reply kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp]: ping

File: kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp (hunk 3 above)
Comment (philbucher) [reply kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp]: ping

File: kratos/utilities/nurbs_utilities/projection_nurbs_geometry_utilities.h [hunk 7]
```diff
@@ -124,15 +124,14 @@ namespace Kratos
...
         // Initialize variables
-        bool clampU = false;
-        bool clampV = false;
+        bool is_first_row_zero, is_second_row_zero, is_first_column_zero, is_second_column_zero, is_system_invertible;
         double d_u = 0.0;
         double d_v = 0.0;
-        double A1_length, A2_length, A1_times_distance_vector, A1_length_times_distance, 
-            A2_times_distance_vector, A2_length_times_distance, A1_times_distance_normalized, A2_times_distance_normalized;
-        double residualU, residualV, J_00, J_01, J_11;
+        double xiCos, etaCos, residualU, residualV, J_00, J_01, J_11, det_J;
```
Comment (philbucher) [inline kratos/utilities/nurbs_utilities/projection_nurbs_geometry_utilities.h]: please make the names lower-case

File: kratos/tests/cpp_tests/geometries/test_projection_nurbs_geometry.cpp (hunk 2 above)
Comment (andreas-apostolatos) [reply kratos/tests/cpp_tests/geometries/test_projection_nurbs_geometry.cpp:191]: yes, I just moved the point closer but not on the singularity

**Comment by andreas-apostolatos** [comment]
Hello together,

I think the issues raised are appropriately addressed, can we please proceed in approving the PR? It seems meanwhile that continuous-integration/appveyor/pr is not running, I restarted yesterday and it timed out again.. Any ideas why?

Thanks!

Best Regards,

File: kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            // Dummy knot vector
+            Vector knot_vector = ZeroVector(number_of_knots);
+            knot_vector[0] = 0.0;
+            knot_vector[1] = 1.0;
+
+            // Dummy curve parameter
+            double parameter_t = 0.5;
+
+            // Call the functions from the geometries/nurbs_shape_function_utilities/nurbs_utilities.h to remove the warning
+            NurbsUtilities::GetNumberOfSpans(polynomial_degree, number_of_knots);
```
Comment (tteschemacher) [inline kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp]: wouldn't it make sense to also test against a value?

Comment (andreas-apostolatos) [reply kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp]: @philbucher suggested to just call the functions in order to skip the warning, but now that you mention that it does really make sense :) thanks!

Comment (tteschemacher) [reply kratos/tests/cpp_tests/geometries/test_removing_warnings_nurbs_utilities.cpp]: Could be just a simple wrapper around it. But, then at least it is really a test. ;) Thanks!

File: kratos/tests/cpp_tests/geometries/test_nurbs_utilities.cpp [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            // Call the NURBS utility functions
+            int number_of_spans = NurbsUtilities::GetNumberOfSpans(polynomial_degree, number_of_knots);
+            polynomial_degree = NurbsUtilities::GetPolynomialDegree(number_of_knots, number_of_control_points);
+            number_of_control_points = NurbsUtilities::GetNumberOfControlPoints(polynomial_degree, number_of_knots);
+            int lower_span = NurbsUtilities::GetLowerSpan(polynomial_degree, knot_vector, parameter_t);
+            int upper_span = NurbsUtilities::GetUpperSpan(polynomial_degree, knot_vector, parameter_t);
+            number_of_knots = NurbsUtilities::GetNumberOfKnots(polynomial_degree, number_of_control_points);
+
+            // Verify the results
+            KRATOS_CHECK_NEAR(number_of_spans, 1, TOLERANCE);
```
Comment (philbucher) [inline kratos/tests/cpp_tests/geometries/test_nurbs_utilities.cpp]: please use `KRATOS_CHECK_EQUAL` for ints

```suggestion
            KRATOS_CHECK_EQUAL(number_of_spans, 1);
```

**Review by philbucher — COMMENTED** [review]
one last comment abt the check-macros, then I will approve

**Review by philbucher — APPROVED** [review]
:+1:

**Pull request merged**